test: Rework ci (Use travis only as fallback env)

maflcko opened the pull request:
This moves the `.travis` folder to `ci` and removes dependence on travis, so that the test script can be run anywhere.

Commit:
ci: Rename .travis/ to ./ci/

Commit:
ci: Remove dependence on travis, use it as fallback env

practicalswift commented:
@MarcoFalke Could you provide some additional context? With this PR merged tests would still run in both Appveyor and Travis, right? What is the plan? :-)

MarcoFalke commented:
The plan is to be flexible, to be able to switch away from travis any time without much hassle. Also, this makes it easier to run the ci locally.

sipa commented on ci/README.md:
```diff
@@ -0,0 +1,8 @@
+
+This directory contains scripts for each build step in each build stage.
+
+Currently three stages `lint`, `extended_lint` and `test` are defined. Each stage has it's own lifecycle, similar to the
```
its

promag commented:
Concept ACK.

Commit:
ci: Add retry

From: https://github.com/kadwanev/retry/tree/a1b1826bdb0a78189d5c70c858dc676f5133b1d7

MarcoFalke commented:
Fixed typo in doc and added a note on how to run the ci locally

Commit:
ci: Use ./ci/ on non-travis host

fanquake reviewed (commented):
Concept ACK. 

Have tested running the CI locally on macOS using Docker and `./ci/test_run_all.sh`. Everything seems to work as expected.

It's a little annoying that we have to bring more bash scripts into this repo, but I guess they can just sit, and we wont have to modify/touch them at all.

edit: Could also add `out/` and `qa-assets/` to `.gitignore`.

MarcoFalke commented on ci/README.md:
````diff
@@ -0,0 +1,24 @@
+requires `docker` to be installed. To install all requirements on Ubuntu, run
+
+```
+sudo apt install docker.io ccache bash git
````
Huh? According to https://packages.ubuntu.com/xenial/docker.io it is the latest version. We don't need any fancy new features (other than `--mount`), so it should be fine as is.

fanquake replied:
Interesting. From the [Docker documentation](https://docs.docker.com/install/linux/docker-ce/ubuntu/):
> Older versions of Docker were called docker, docker.io , or docker-engine. If these are installed, uninstall them:
> ```bash
> sudo apt-get remove docker docker-engine docker.io containerd runc
> ```

I guess I don't mind either way, just wasn't sure after looking at those docs.

MarcoFalke commented:
> Have tested running the CI locally on macOS using Docker and ./ci/test_run_all.sh. Everything seems to work as expected.

Thanks for testing on mac!

So this is confirmed to be working on:
* Locally, macOS host (fanquake)
* travis, xenial host (run as part of this pull)
* cirrus ci, bionic host (https://cirrus-ci.com/task/5077011800784896)
* GitHub ci, bionic host (https://github.com/MarcoFalke/bitcoin/runs/190746548)
* Locally, bionic host (MarcoFalke)

instagibbs commented:
> Also, this makes it easier to run the ci locally.

Concept ACK!

practicalswift commented on ci/retry/retry:
```diff
@@ -0,0 +1,163 @@
+  echo "$1" 1>&2
+}
+
+# Paramters: max_tries min_sleep max_sleep constant_sleep fail_script EXECUTION_COMMAND
```
Should be "parameters" :-)

fanquake replied:
This is an [upstream script](https://github.com/kadwanev/retry), so we don't need to fix it's issues here. Regardless, I've [PR'd](https://github.com/kadwanev/retry/pull/15) them upstream.

practicalswift commented on ci/retry/retry:
```diff
@@ -0,0 +1,163 @@
+    -v, --verbose                    Verbose output
+    -t, --tries=#                    Set max retries: Default 10
+    -s, --sleep=secs                 Constant sleep amount (seconds)
+    -m, --min=secs                   Exponenetial Backoff: minimum sleep amount (seconds): Default 0.3
```
Should be exponential -- applies throughout PR :-)

practicalswift commented on .cirrus.yml:
```diff
@@ -27,3 +27,26 @@ task:
     - gmake check ${MAKEJOBS} VERBOSE=1
   functional_test_script:
     - ./test/functional/test_runner.py --jobs 9 --ci --extended --exclude feature_dbcrash --combinedlogslen=1000 --quiet --failfast
+task:
```
Duplication of key `task` in mapping?

MarcoFalke replied:
Yes, this is the correct syntax:

https://cirrus-ci.org/guide/writing-tasks/#task-execution-dependencies

practicalswift commented on .cirrus.yml:
```diff
@@ -27,3 +27,26 @@ task:
+    folder: "/tmp/ccache_dir"
+  depends_cache:
+    folder: "/tmp/cirrus-ci-build/depends/built"
+    folder: "/tmp/cirrus-ci-build/depends/sdk-sources"
```
Duplication of key `folder` in mapping?

MarcoFalke commented:
Rebased with three-way-merge due to change in tangent line

fanquake commented:
I'm now seeing an error running this:
```
configure.ac:40: installing 'build-aux/missing'
src/Makefile.am: installing 'build-aux/depcomp'
travis_fold:end:autogen

travis_fold:start:configure
configure: loading site script /Users/michael/github/bitcoin/depends/x86_64-unknown-linux-gnu/share/config.site
configure: loading cache config.cache
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for x86_64-unknown-linux-gnu-strip... no
checking for strip... strip
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... no
checking for mawk... mawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
configure: error: source directory already configured; run "make distclean" there first
This file contains any messages produced by compilers while
running configure, to aid debugging if configure makes a mistake.

It was created by Bitcoin Core configure 0.18.99, which was
generated by GNU Autoconf 2.69.  Invocation command line was

  $ ../configure --cache-file=config.cache --disable-dependency-tracking --prefix=/Users/michael/github/bitcoin/depends/x86_64-unknown-linux-gnu --bindir=/Users/michael/github/bitcoin/out/x86_64-unknown-linux-gnu/bin --libdir=/Users/michael/github/bitcoin/out/x86_64-unknown-linux-gnu/lib
```

MarcoFalke commented:
It says `source directory already configured; run "make distclean" there first`. Have you tried that?

MarcoFalke commented:
Specifically `cd /Users/michael/github/bitcoin && make distclean`.

The ci is generally assuming the the folder is a fresh git clone. If needed, that can be changed later on, but better not part of this pull.

fanquake commented:
> It says source directory already configured; run "make distclean" there first. Have you tried that?

Shouldn't that all be handled by the script? If not, then some docs about why you can't run the tests twice in a row would be handy.

MarcoFalke commented:
Done

Commit:
ci: Add note that this assumes a fresh git clone

fanquake reviewed (approved):
Thanks ACK fa9404d76d16c618dbd2c19973246900918f9e6c - Maybe @sipsorcery or @NicolasDorier might also want to test here, then we'd have all OS's covered?

MarcoFalke commented:
I don't think docker on windows can run Ubuntu containers

MarcoFalke commented:
Rebased without any changes due to overly sensitive merge conflict detection on GitHub. git range-diff should be empty: `git range-diff bitcoin/master fa9404d76d faeacf3269`

sipsorcery commented:
> I don't think docker on windows can run Ubuntu containers

It can since https://docs.docker.com/docker-for-windows/wsl-tech-preview/

````
aaron@pcdodo:~$ uname -a
Linux pcdodo 4.19.57-microsoft-standard #1 SMP Wed Jul 3 20:47:25 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
aaron@pcdodo:~$ docker run ubuntu echo "I  love Windows"; lsb_release -a
I  love Windows
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.2 LTS
Release:        18.04
Codename:       bionic
````
Actually it was already possible to use ubuntu docker containers on Windows but now it  possible directly on top of the Windows Subsystem for Linux kernel.

sipsorcery commented:
> Thanks ACK [fa9404d](https://github.com/bitcoin/bitcoin/commit/fa9404d76d16c618dbd2c19973246900918f9e6c) - Maybe @sipsorcery or @NicolasDorier might also want to test here, then we'd have all OS's covered?

@fanquake I looked over the PR and successfully ran my appveyor build with the PR merged. Doesn't seem to be anything Windows related in this PR? Ony travis and shell scripts. Or did you testing the script changes with the WSL/Mingw builds on Windows?

fanquake commented:
Thanks @sipsorcery. I also thought it was possible to run on Windows.

> Or did you testing the script changes with the WSL/Mingw builds on Windows?

Yea I just meant checking out this PR, and testing/running the scripts, to make sure they were working in your Windows development setups.
